src/providers/ipa/ipa_hbac_common.c

```diff
@@ -533,56 +536,108 @@ hbac_eval_user_element(TALLOC_CTX *mem_ctx,
     }
     users->name = talloc_steal(users, shortname);
 
-    /* Read the originalMemberOf attribute
-     * This will give us the list of both POSIX and
-     * non-POSIX groups that this user belongs to.
-     */
+    ipa_domain = get_domains_head(domain);
+    if (ipa_domain == NULL) {
+        ret = EINVAL;
+        goto done;
+    }
+
+    ipa_groups_basedn = ldb_dn_new_fmt(tmp_ctx, sysdb_ctx_get_ldb(domain->sysdb),
+                                        SYSDB_TMPL_GROUP_BASE, ipa_domain->name);
+    if (ipa_groups_basedn == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
     ret = sysdb_search_user_by_name(tmp_ctx, domain, username,
-                                    attrs, &msg);
+                                    NULL, &msg);
     if (ret != EOK) {
         DEBUG(SSSDBG_CRIT_FAILURE,
               "Could not determine user memberships for [%s]\n",
                   users->name);
         goto done;
     }
 
-    el = ldb_msg_find_element(msg, SYSDB_ORIG_MEMBEROF);
-    if (el == NULL || el->num_values == 0) {
+    /*
+     * Get the name attribute of all groups pointed to by the memberof
+     * attribute. This includes both POSIX and non-POSIX groups.
+     */
+    ret = sysdb_asq_search(tmp_ctx, domain, msg->dn,
```

Comment:
I wonder if instead calling sysdb_search_user_by_name() and sysdb_asq_search() it might be better to just call sysdb_initgroups(). This will return more attributes as we need here but would have the advantage that the very same call is used by the client side initgroups calls, so we should always get the same results as a 'id' call from the shell as long as there are no changes in the cache.